youtube_dl/extractor/xiami.py

```diff
@@ -0,0 +1,135 @@
+                'album': '悟空',
+                'description': 'md5:206e67e84f9bed1d473d04196a00b990',
+            }
+        },
+        {
+            'url': 'http://www.xiami.com/album/2100300444',
+            'info_dict': {
+                'id': '2100300444',
+            },
+            'playlist_count': 10,
+        },
+        {
+            'url': 'http://www.xiami.com/album/512288?spm=a1z1s.6843761.1110925389.6.hhE9p9',
+            'only_matching': True,
+        }
+    ]
+
+    _TYPES = {
+        'song': '',
+        'album': '/type/1',
+        'artist': '/type/2',
+        'collect': '/type/3',
+    }
+
+    _XML_BASE_URL = 'http://www.xiami.com/song/playlist/id'
+    _XML_NS = '{http://xspf.org/ns/0/}'
+
+    def _real_extract(self, url):
+        match = re.match(self._VALID_URL, url)
+        assert(match)
+        typ, _id = match.group('typ'), match.group('id')
+
+        if typ not in self._TYPES:
+            raise ExtractorError('URL not supported')
+        xml_url = '%s/%s' % (self._XML_BASE_URL, _id)
+        xml_url += self._TYPES[typ]
+
+        playlist = self._download_xml(xml_url, _id)
+        tracklist = xpath_element(playlist, './' + self._XML_NS + 'trackList')
+
+        def _extract_single(track):
+            artist = xpath_text(track, self._XML_NS + 'artist', default='')
+            artist = artist.split(';')
+
+            ret = {
+                'id': xpath_text(track, self._XML_NS + 'song_id'),
+                'title': xpath_text(track, self._XML_NS + 'title'),
+                'album': xpath_text(track, self._XML_NS + 'album_name'),
+                'artist': ';'.join(artist) if artist else None,
+                'creator': artist[0] if artist else None,
+                'url': self._decrypt(xpath_text(track, self._XML_NS + 'location')),
+                'thumbnail': xpath_text(track, self._XML_NS + 'pic', default=None),
+                'duration': int(xpath_text(track, self._XML_NS + 'length')),
+            }
+
+            lyrics_url = xpath_text(track, self._XML_NS + 'lyric')
+            if lyrics_url and lyrics_url.endswith('.lrc'):
+                ret['description'] = self._download_webpage(lyrics_url, ret['id'])
+            return ret
+
+        result = [_extract_single(track) for track in tracklist]
+        if not result:
+            raise ExtractorError('No track found')
+        if typ == 'song':
+            result = result[0]
+        else:
+            result = {
+                "_type": "playlist",
+                "entries": result,
+                'id': _id
+            }
+        return result
+
+    @staticmethod
+    def _decrypt(origin):
+        n = int(origin[0])
+        origin = origin[1:]
+        short_lenth = len(origin) / n
+        long_num = len(origin) - short_lenth * n
+        l = tuple()
+        for i in range(0, n):
+            length = short_lenth
+            if i < long_num:
+                length += 1
+            l += (origin[0:length], )
```

Comment:
It seems like to be geo restricted (only available in mainland china). Should I raise an error for this?

(from VPS at USA:)

```
> curl http://www.xiami.com/song/playlist/id/1771975656 | grep available
                                <p>Xiami is currently not available in your country.</p>
```